### PR TITLE
xdg_shell, multi-output, remove sprawl, cleanup

### PR DESCRIPTION
**Pull request by jwinarske**
xdg_shell
-replaces wl_shell for xdg_shell

Multiple outputs
-Adds support for saving state on multiple outputs
-Enables deriving the buffer scaling factor

Cleanup
-Remove unused methods/variables
-Remove unused headers
-Initialize uninitialize variables
-rename m_engine to m_flutter_engine in App class

Remove sprawl option
-Not required with xdg_shell

Update README.md
-Enabling flutter desktop causes App build issue.
 This has been the case for some time.
 Only use custom devices for debugging eLinux Flutter.

FPS feature cleanup
-Fix narrowing warnings
-Update README for usage

Remove pinch gesture references
-Dart layer gesture detection being used.
 Not all targets have this protocol

### CODE REVIEW

File: shell/egl_window.cc [hunk 1]
```diff
@@ -52,18 +51,23 @@ EglWindow::EglWindow(size_t index,
...
+      xdg_wm_base_get_xdg_surface(m_display->GetXdgWmBase(), m_surface);
+  assert(m_xdg_surface);
+  FML_DLOG(INFO) << "EglWindow::m_xdg_surface = "
+                 << static_cast<void*>(m_xdg_surface);
+
+  xdg_surface_add_listener(m_xdg_surface, &xdg_surface_listener, this);
+  //  xdg_surface_add_listener(m_xdg_surface, &xdg_surface_listener,this);
+  m_xdg_toplevel = xdg_surface_get_toplevel(m_xdg_surface);
+  xdg_toplevel_set_title(m_xdg_toplevel, m_app_id.c_str());
+  //  wl_surface_commit(m_surface);
```
Comment (mv0) [inline shell/egl_window.cc]: Better to just remove it entirely.

File: shell/display.cc [hunk 2]
```diff
@@ -170,7 +170,7 @@ void Display::registry_handle_global(
 
   else if (strcmp(interface, wl_output_interface.name) == 0) {
     d->m_output = static_cast<struct wl_output*>(
-        wl_registry_bind(registry, name, &wl_output_interface, 1));
+        wl_registry_bind(registry, name, &wl_output_interface, 2));
```
Comment (mv0) [inline shell/display.cc]: For the versioning we usually do MIN(2, version).

File: shell/egl_window.cc [hunk 3]
```diff
@@ -107,9 +107,9 @@ EglWindow::EglWindow(size_t index,
   m_fps_idx = 0;
   m_fps_counter = 0;
 
-  //  this->m_callback = wl_surface_frame(this->m_surface);
-  //  wl_callback_add_listener(this->m_callback, &frame_listener, this);
-  wl_surface_commit(this->m_surface);
+  m_callback = wl_surface_frame(this->m_surface);
+  wl_callback_add_listener(m_callback, &frame_listener, this);
+  wl_surface_commit(m_surface);
```
Comment (mv0) [inline shell/egl_window.cc]: Probably this wouldn't need any comments from the previous patch. Also wouldn't `this->m_surface == m_surface`?

File: shell/egl_window.cc [hunk 4]
```diff
@@ -336,6 +336,9 @@ void EglWindow::redraw(void* data,
   if (callback)
     wl_callback_destroy(callback);
 
+  window->m_callback = wl_surface_frame(window->m_surface);
+  wl_callback_add_listener(window->m_callback, &frame_listener, window);
```
Comment (mv0) [inline shell/egl_window.cc:373]: I think this belongs in a different patch. Obviously not adding a new frame callback would not trigger it a new one. So this seems like a small bug missing adding it. Right?

File: shell/egl_window.cc [hunk 5]
```diff
@@ -361,9 +364,9 @@ void EglWindow::toggle_fullscreen() {
     xdg_toplevel_unset_fullscreen(m_xdg_toplevel);
   }
 
-  //handle_xdg_surface_configure(this, m_xdg_surface, 0);
-//  wl_callback_add_listener(wl_display_sync(m_display->GetDisplay()),
-//                           &xdg_configure_callback_listener, this);
+  // handle_xdg_surface_configure(this, m_xdg_surface, 0);
+  //  wl_callback_add_listener(wl_display_sync(m_display->GetDisplay()),
+  //                           &xdg_configure_callback_listener, this);
```
Comment (mv0) [inline shell/egl_window.cc]: Again, this probably is just indenting thing.

File: shell/egl_window.h [hunk 6]
```diff
@@ -112,32 +113,10 @@ class EglWindow : public Egl {
...
-                                  int width,
-                                  int height,
-                                  uint32_t time);
-
-  static void paint_pixels(void* image,
-                           int padding,
-                           int width,
-                           int height,
-                           uint32_t time);
-
```
Comment (mv0) [inline shell/egl_window.h:151]: Clean-ups of these do belong in a different patch. These removal have nothing to do with patch. Or if so please why in the commit.

File: shell/egl_window.cc [hunk 7]
```diff
@@ -49,8 +49,9 @@ EglWindow::EglWindow(size_t index,
   FML_DLOG(INFO) << "+ EglWindow()";
 
   m_surface = wl_compositor_create_surface(m_display->GetCompositor());
-  assert(m_surface);
-  FML_DLOG(INFO) << "EglWindow::m_surface = " << static_cast<void*>(m_surface);
+#if HIDPI_CLIENT
```
Comment (mv0) [inline shell/egl_window.cc]: Why only for HDPI_CLIENT? 

What's the reasoning. There's no commit description explaining it.

File: shell/egl_window.cc [hunk 8]
```diff
@@ -49,8 +49,9 @@ EglWindow::EglWindow(size_t index,
   FML_DLOG(INFO) << "+ EglWindow()";
 
   m_surface = wl_compositor_create_surface(m_display->GetCompositor());
-  assert(m_surface);
-  FML_DLOG(INFO) << "EglWindow::m_surface = " << static_cast<void*>(m_surface);
```
Comment (mv0) [inline shell/egl_window.cc:54]: Removal of these changes require a different patch.

File: shell/egl_window.cc [hunk 9]
```diff
@@ -59,15 +60,9 @@ EglWindow::EglWindow(size_t index,
 
   m_xdg_surface =
       xdg_wm_base_get_xdg_surface(m_display->GetXdgWmBase(), m_surface);
-  assert(m_xdg_surface);
-  FML_DLOG(INFO) << "EglWindow::m_xdg_surface = "
-                 << static_cast<void*>(m_xdg_surface);
```
Comment (mv0) [inline shell/egl_window.cc]: Same. Removal of stuff require a different patch.

File: shell/egl_window.cc [hunk 10]
```diff
@@ -59,15 +60,9 @@ EglWindow::EglWindow(size_t index,
...
-
   xdg_surface_add_listener(m_xdg_surface, &xdg_surface_listener, this);
-  //  xdg_surface_add_listener(m_xdg_surface, &xdg_surface_listener,this);
```
Comment (mv0) [inline shell/egl_window.cc]: This belongs in a previous patch.

File: shell/egl_window.cc [hunk 11]
```diff
@@ -59,15 +60,9 @@ EglWindow::EglWindow(size_t index,
...
   m_xdg_toplevel = xdg_surface_get_toplevel(m_xdg_surface);
   xdg_toplevel_set_title(m_xdg_toplevel, m_app_id.c_str());
-  //  wl_surface_commit(m_surface);
```
Comment (mv0) [inline shell/egl_window.cc]: Belongs in a different/previous patch.

File: shell/egl_window.cc [hunk 12]
```diff
@@ -84,18 +79,8 @@ EglWindow::EglWindow(size_t index,
 
   m_egl_surface[index] = create_egl_surface(this, m_egl_window[index], nullptr);
 
-  toggle_fullscreen();
```
Comment (mv0) [inline shell/egl_window.cc:85]: Yeah, I've already made a comment about this, but reviewing happens patch by patch. I would introduce these changes in first patch probably.

File: shell/egl_window.cc [hunk 13]
```diff
@@ -325,6 +310,30 @@ void EglWindow::handle_xdg_surface_configure(void* data,
   w->m_xdg_surface_is_configured = true;
 }
 
+#if HIDPI_CLIENT
```
Comment (mv0) [inline shell/egl_window.cc]: Again, no explanation as to why only for HIDPI_CLIENT.

File: shell/display.cc [hunk 14]
```diff
@@ -741,6 +715,53 @@ bool Display::ActivateSystemCursor([[maybe_unused]] int32_t device,
...
+    }
+  }
+}
+
+const struct wl_surface_listener Display::primary_surface_listener = {
+    handle_primary_surface_enter};
+
+void Display::SetSurface(wl_surface* surface) {
+  m_surface = surface;
+  // requires compositor v3+
```
Comment (mv0) [inline shell/display.cc:764]: What is compositor v3?

File: shell/display.cc [hunk 15]
```diff
@@ -741,6 +715,53 @@ bool Display::ActivateSystemCursor([[maybe_unused]] int32_t device,
...
+  // wl_surface_set_buffer_scale(m_surface, m_buffer_scale);
```
Comment (mv0) [inline shell/display.cc:765]: Left over?

File: shell/display.h [hunk 16]
```diff
@@ -177,33 +166,24 @@ class Display {
...
+    struct wl_output* output;
+    uint32_t global_id;
+    unsigned width;
+    unsigned height;
+    unsigned physical_width;
+    unsigned physical_height;
+    int refresh_rate;
+    int32_t scale;
+    bool done;
+  } output_info_t;
```
Comment (mv0) [inline shell/display.h:179]: So... .wouldn't this better be a class? Instead of typedefing stuff. Its CPP after all.

File: shell/egl_window.cc [hunk 17]
```diff
@@ -62,39 +56,29 @@ EglWindow::EglWindow(size_t index,
       xdg_wm_base_get_xdg_surface(m_display->GetXdgWmBase(), m_surface);
   xdg_surface_add_listener(m_xdg_surface, &xdg_surface_listener, this);
   m_xdg_toplevel = xdg_surface_get_toplevel(m_xdg_surface);
+  xdg_toplevel_add_listener(m_xdg_toplevel, &xdg_toplevel_listener, this);
```
Comment (mv0) [inline shell/egl_window.cc:59]: This needs to be in the first patch.

File: shell/egl_window.cc [hunk 18]
```diff
@@ -305,37 +289,77 @@ int EglWindow::create_shm_buffer(Display* display,
...
+  if (width > 0 && height > 0) {
+    if (!w->m_fullscreen && !w->m_maximized) {
+      w->m_window_size.width = width;
+      w->m_window_size.height = height;
+    }
+    w->m_geometry.width = width;
+    w->m_geometry.height = height;
+
+    /* Initialise damage to full surface, so the padding gets painted */
+    wl_surface_damage(w->m_surface, 0, 0, w->m_geometry.width,
```
Comment (mv0) [inline shell/egl_window.cc:337]: No need to inflict damage for the surface.

File: shell/egl_window.cc [hunk 19]
```diff
@@ -305,37 +289,77 @@ int EglWindow::create_shm_buffer(Display* display,
...
+                      w->m_geometry.height);
+
+    if (w->m_egl_window[w->m_index]) {
```
Comment (mv0) [inline shell/egl_window.cc:340]: I would do the resize at the end.

File: shell/egl_window.cc [hunk 20]
```diff
@@ -366,16 +390,8 @@ const struct wl_callback_listener EglWindow::frame_listener = {redraw};
...
-    xdg_toplevel_set_fullscreen(m_xdg_toplevel, m_display->GetOutput());
-  } else {
-    xdg_toplevel_unset_fullscreen(m_xdg_toplevel);
-  }
-
-  // handle_xdg_surface_configure(this, m_xdg_surface, 0);
-  //  wl_callback_add_listener(wl_display_sync(m_display->GetDisplay()),
-  //                           &xdg_configure_callback_listener, this);
+void EglWindow::set_fullscreen() {
+  xdg_toplevel_set_fullscreen(m_xdg_toplevel, nullptr);
```
Comment (mv0) [inline shell/egl_window.cc:394]: You've changed this 3 times in this series. I would keep the bool as is. You might want to change from fullscreen to unfulscreen on different shells.

File: shell/egl_window.h [hunk 21]
```diff
@@ -27,7 +27,13 @@
 #include "egl.h"
 #include "xdg-shell-client-protocol.h"
 
+#define WL_ARRAY_FOR_EACH(pos, array, type)                             \
```
Comment (mv0) [inline shell/egl_window.h:30]: #ifndef ....

Please also explain why we need this.... I'd like to see some cpp construct that fixes this... seems a bit weird that can't really cast generic pointers. In C we don't even need an explicit cast.

File: shell/egl_window.h [hunk 22]
```diff
@@ -67,22 +72,30 @@ class EglWindow : public Egl {
...
   bool m_fullscreen;
-  bool m_sprawl;
+  bool m_maximized{};
+  bool m_resize{};
+  bool m_activated{};
+  bool m_running{};
+  struct {
+    int32_t width;
+    int32_t height;
+  } m_geometry;
```
Comment (mv0) [inline shell/egl_window.h:88]: `m_geometry, m_window_size{}`?

File: shell/egl_window.h [hunk 23]
```diff
@@ -67,22 +72,30 @@ class EglWindow : public Egl {
...
+  struct {
+    int32_t width;
+    int32_t height;
+  } m_window_size{};
+
```
Comment (mv0) [inline shell/egl_window.h:93]: No need do declare once more, there's one above.

File: shell/egl_window.h [hunk 24]
```diff
@@ -94,7 +107,7 @@ class EglWindow : public Egl {
 
   int m_frame_sync;
 
-  void set_fullscreen(bool val);
+  void set_fullscreen();
```
Comment (mv0) [inline shell/egl_window.h:110]: Wasn't okay to able to go back from fullscreen to unfullscreen?

File: shell/egl_window.h [hunk 25]
```diff
@@ -107,25 +120,22 @@ class EglWindow : public Egl {
...
+                                        struct xdg_toplevel* toplevel,
+                                        int32_t width,
+                                        int32_t height,
+                                        struct wl_array* states);
 
-  static void handle_primary_surface_leave(void* data,
-                                           struct wl_surface* wl_surface,
-                                           struct wl_output* output);
-#endif
+  static void handle_toplevel_close(void* data,
```
Comment (mv0) [inline shell/egl_window.h:137]: All of the toplevel bits need to go in the first patch, added here need to go in the first patch.

File: shell/engine.h [hunk 26]
```diff
@@ -122,9 +122,9 @@ class Engine {
   std::string GetAssetDirectory() { return m_assets_path; }
 
 #if ENABLE_PLUGIN_TEXT_INPUT
-  TextInput* m_text_input;
+  TextInput* m_text_input{};
   [[maybe_unused]] void SetTextInput(TextInput *text_input);
-  [[maybe_unused]] TextInput *GetTextInput();
+  [[maybe_unused]] TextInput *GetTextInput() const;
```
Comment (mv0) [inline shell/engine.h:127]: These belong in a different patch.

File: shell/egl_window.cc [hunk 27]
```diff
@@ -62,39 +56,29 @@ EglWindow::EglWindow(size_t index,
...
-  FML_DLOG(INFO) << "create egl_window: " << m_width << "x" << m_height;
-
-  m_egl_surface[index] = create_egl_surface(this, m_egl_window[index], nullptr);
-
-  set_fullscreen(m_fullscreen);
-
-  /* Initialise damage to full surface, so the padding gets painted */
-  wl_surface_damage(m_surface, 0, 0, m_width, m_height);
+  m_egl_window[m_index] =
+      wl_egl_window_create(m_surface, m_geometry.width, m_geometry.height);
```
Comment (mv0) [inline shell/egl_window.cc:66]: The m_geometry.width and m_geometry.height are passed from App() right? 

In practice, are these the updated values from the resize  in top level configure or the initial ones?

Have you tried these in system with software rendering?

**Comment by mv0** [comment]
My initial comments for these are:

- please keep the cleanups in a different patch/patches
- keep together the wl_shell to xdg-shell conversion - do not split toplevel configure events from xdg_surface ones. Makes reviewing much easier
- all patches are missing a proper commit description

File: shell/egl_window.cc (hunk 27 above)
Comment (jwinarske) [reply shell/egl_window.cc:66]: The initial values passed from App() get updated via handle_toplevel_configure().

I have not yet tried this PR with a software render (Qemu).  I need to validate it on my CI platforms (meta-flutter).

I am planning at some point to add optional support for a decoration library for a more friendly desktop support.  So resizing etc will be added later.  The primary use case is a minimal target app, that happens to work on desktop.

File: shell/display.h [hunk 28]
```diff
@@ -237,7 +238,7 @@ class Display {
                                    int scale);
   static void display_handle_done(void* data, struct wl_output* wl_output);
 
-  bool m_is_configured;
+  bool m_is_configured{};
```
[the remark this answers is not in the document]
Comment (jwinarske) [reply shell/display.h:222]: Yes.

File: shell/egl_window.cc [hunk 29]
```diff
@@ -80,21 +84,17 @@ EglWindow::EglWindow(size_t index,
 
   m_egl_surface[index] = create_egl_surface(this, m_egl_window[index], nullptr);
 
-  wl_shell_surface_set_title(m_shell_surface, m_app_id.c_str());
-
   toggle_fullscreen();
```
[the remark this answers is not in the document]
Comment (jwinarske) [reply shell/egl_window.cc]: Moved this to !m_fullscreen case.

File: shell/egl_window.cc [hunk 30]
```diff
@@ -317,122 +317,16 @@ int EglWindow::create_shm_buffer(Display* display,
...
-
-void EglWindow::paint_pixels_bottom(void* image,
-                                    [[maybe_unused]] int padding,
-                                    int width,
-                                    int height,
-                                    [[maybe_unused]] uint32_t time) {
-  memset(image, 0xa0,
-         static_cast<size_t>(width) * static_cast<size_t>(height) * 4);
+  xdg_surface_ack_configure(xdg_surface, serial);
+  w->m_xdg_surface_is_configured = true;
```
[the remark this answers is not in the document]
Comment (jwinarske) [reply shell/egl_window.cc]: The wl_callback_add_listener is used for painting the FPS window.

File: shell/egl_window.h [hunk 31]
```diff
@@ -72,7 +69,7 @@ class EglWindow : public Egl {
   std::shared_ptr<Display> m_display;
 
   struct wl_surface* m_surface;
-  struct wl_shell_surface* m_shell_surface;
+  struct wl_shell_surface* m_shell_surface{};
```
[the remark this answers is not in the document]
Comment (jwinarske) [reply shell/egl_window.h]: yep

File: shell/egl_window.cc (hunk 3 above)
Comment (jwinarske) [reply shell/egl_window.cc]: yes, left over from prior patch.  remove

File: shell/egl_window.h (hunk 21 above)
Comment (jwinarske) [reply shell/egl_window.h:31]: The base Wayland macro does not compile as-is in C++.  This is pretty well documented on the internet.  I added a comment in updated patches.

File: shell/egl_window.h (hunk 22 above)
Comment (jwinarske) [reply shell/egl_window.h:89]: Optional window decorations are next hence window vs geometry.

File: shell/egl_window.cc (hunk 18 above)
Comment (jwinarske) [reply shell/egl_window.cc:341]: removed

File: shell/egl_window.cc (hunk 4 above)
Comment (jwinarske) [reply shell/egl_window.cc]: this is for rendering the FPS overlay window

File: shell/display.cc (hunk 2 above)
Comment (jwinarske) [reply shell/display.cc:168]: noted and addressed

File: shell/display.h (hunk 16 above)
Comment (jwinarske) [reply shell/display.h:179]: If you search the source tree you'll find many references to this pattern.  The goal is to stay as close to Google/Flutter standards.  This makes upstreaming much easier.  So this is fine.

File: shell/egl_window.cc (hunk 19 above)
Comment (jwinarske) [reply shell/egl_window.cc:341]: moved.

File: shell/egl_window.cc [hunk 32]
```diff
@@ -52,18 +51,23 @@ EglWindow::EglWindow(size_t index,
...
+  m_xdg_surface =
+      xdg_wm_base_get_xdg_surface(m_display->GetXdgWmBase(), m_surface);
+  assert(m_xdg_surface);
+  FML_DLOG(INFO) << "EglWindow::m_xdg_surface = "
+                 << static_cast<void*>(m_xdg_surface);
+
+  xdg_surface_add_listener(m_xdg_surface, &xdg_surface_listener, this);
+  //  xdg_surface_add_listener(m_xdg_surface, &xdg_surface_listener,this);
+  m_xdg_toplevel = xdg_surface_get_toplevel(m_xdg_surface);
+  xdg_toplevel_set_title(m_xdg_toplevel, m_app_id.c_str());
```
[the remark this answers is not in the document]
Comment (jwinarske) [reply shell/egl_window.cc]: done.

**Comment by jwinarske** [comment]
Yeah I agree.  This wasn't ready for review when I pointed it out.  I refactored the patches for logical grouping.

File: shell/egl_window.h (hunk 21 above)
Comment (jwinarske) [reply shell/egl_window.h:31]: https://gitlab.freedesktop.org/wayland/wayland/-/issues/34

**Review by cmc5788 — DISMISSED** [review]

File: shell/egl_window.h (hunk 22 above)
Comment (mv0) [reply shell/egl_window.h:89]: Yeah, I was just hinting that we can use the same struct  here instead of declaring a new anonymous one.

**Comment by mv0** [comment]
Had a few minor nitpicks, these now look fine to me now. I (still) need to do a follow-up with the resize, as that will currently will be an issue when starting up in a different state (max, rather than full). Feel free to merge once resolved (the comments).

File: README.md [hunk 33]
```diff
@@ -108,6 +108,21 @@ Copy libflutter_engine.so to `/usr/local/lib` or use LD_LIBRARY_PATH to point do
     export LD_LIBRARY_PATH=`pwd`:$LD_LIBRARY_PATH
     homescreen
 
+## FPS
+Three environmental variables impact FPS reporting.
```
[the remark this answers is not in the document]
Comment (jwinarske) [reply README.md:112]: no I haven't.  I'll check that out.

File: shell/display.cc [hunk 34]
```diff
@@ -194,69 +198,55 @@ void Display::display_handle_geometry(void* data,
...
-    FML_DLOG(INFO) << "subpixel: " << d->m_info.geometry.subpixel;
-    FML_DLOG(INFO) << "make: " << d->m_info.geometry.make;
-    FML_DLOG(INFO) << "model: " << d->m_info.geometry.model;
-    FML_DLOG(INFO) << "transform: " << d->m_info.geometry.transform;
-  }
+  auto* oi = static_cast<output_info_t*>(data);
+  oi->physical_width = physical_width;
+  oi->physical_height = physical_height;
+
+  FML_DLOG(INFO) << "Physical width: " << physical_width << " mm x "
```
[the remark this answers is not in the document]
Comment (jwinarske) [reply shell/display.cc:227]: Correct.  Pass `-DCMAKE_BUILD_TYPE=Debug` to CMake.

File: shell/app.cc [hunk 35]
```diff
@@ -107,8 +107,9 @@ App::App(const std::string& app_id,
   m_fps_counter = 0;
 
   const char* envstr_console;
+  char* pEnd;
   if ((envstr_console = getenv("FPS_OUTPUT_CONSOLE")) != nullptr) {
-    int val = atoi(envstr_console);
+    long val = strtol(envstr_console, &pEnd, 10);
```
[the remark this answers is not in the document]
Comment (jwinarske) [reply shell/app.cc:108]: good point.  removed pEnd for nullptr.

File: shell/display.cc [hunk 36]
```diff
@@ -32,15 +33,16 @@ Display::Display([[maybe_unused]] App* app,
...
       m_subcompositor(nullptr),
       m_cursor_surface(nullptr),
       m_keyboard(nullptr),
       m_xkb_context(xkb_context_new(XKB_CONTEXT_NO_FLAGS)),
       m_keymap(nullptr),
       m_xkb_state(nullptr),
-      m_agl_shell(nullptr),
+      m_buffer_scale(1),
+      m_last_buffer_scale(1),
+	  m_agl_shell(nullptr),
```
[the remark this answers is not in the document]
Comment (jwinarske) [reply shell/display.cc]: I've refactored this.  All nullptr initialization is happening at point of definition.

File: shell/app.cc [hunk 37]
```diff
@@ -57,7 +57,7 @@ App::App(const std::string& app_id,
 
   FML_DLOG(INFO) << "+App::App";
 
-  m_display->AglShellDoBackground(m_egl_window[0]->GetNativeSurface());
+  m_display->AglShellDoBackground(m_egl_window[0]->GetNativeSurface(), 0);
```
[the remark this answers is not in the document]
Comment (jwinarske) [reply shell/app.cc:60]: Yeah I agree.  I'll add `constexpr int` values to constants.h once the layout is logically tied to a use case.  For now it's more of a placeholder than anything.

File: shell/egl_window.cc [hunk 38]
```diff
@@ -53,51 +52,35 @@ EglWindow::EglWindow(size_t index,
...
-  this->m_callback = wl_surface_frame(this->m_surface);
-  wl_callback_add_listener(this->m_callback, &frame_listener, this);
-  wl_surface_commit(this->m_surface);
+  m_callback = wl_surface_frame(m_surface);
+  wl_callback_add_listener(m_callback, &frame_listener, this);
+
+  if (m_fullscreen)
+    xdg_toplevel_set_fullscreen(m_xdg_toplevel, nullptr);
+  else
+    xdg_toplevel_set_title(m_xdg_toplevel, m_app_id.c_str());
```
[the remark this answers is not in the document]
Comment (jwinarske) [reply shell/egl_window.cc:83]: I moved it back.

File: shell/display.cc [hunk 39]
```diff
@@ -194,69 +198,55 @@ void Display::display_handle_geometry(void* data,
...
-    FML_DLOG(INFO) << "scale: " << d->m_info.scale.scale;
-  }
+  FML_DLOG(INFO) << "Display Scale Factor: " << factor;
 }
 
 void Display::display_handle_done(void* data, struct wl_output* wl_output) {
-  (void)data;
   (void)wl_output;
+  auto* oi = static_cast<output_info_t*>(data);
+  oi->done = true;
```
[the remark this answers is not in the document]
Comment (jwinarske) [reply shell/display.cc:262]: Hmm.  On multiple platforms I'm seeing done always show up last.  I'm not currently being used for anything.

File: shell/display.h [hunk 40]
```diff
@@ -166,30 +168,24 @@ class Display {
...
+  typedef struct output_info {
+    struct wl_output* output;
+    uint32_t global_id;
+    unsigned width;
+    unsigned height;
+    unsigned physical_width;
+    unsigned physical_height;
+    int refresh_rate;
+    int32_t scale;
+    bool done;
```
[the remark this answers is not in the document]
Comment (jwinarske) [reply shell/display.h:180]: I need to read the spec on this.  I see done in the log as the last wl_output event.  Again I'm not taking action on the value.

File: shell/display.h [hunk 41]
```diff
@@ -166,30 +168,24 @@ class Display {
...
-    } scale{};
-
-  } m_info;
+  typedef struct output_info {
+    struct wl_output* output;
+    uint32_t global_id;
+    unsigned width;
+    unsigned height;
+    unsigned physical_width;
+    unsigned physical_height;
```
[the remark this answers is not in the document]
Comment (jwinarske) [reply shell/display.h:177]: Yeah I'm not using the EDID description at this time.  I may add this back to pass to upper level app at some point.  There are other things that need to be reported as well.  Not a firm use case yet.

File: shell/display.cc (hunk 36 above)
[the remark this answers is not in the document]
Comment (jwinarske) [reply shell/display.cc]: not sure what you're referring to here.

**Comment by jwinarske** [comment]
@mv0 I have updates that handle buffering scaling, and setup interface versioning using std::min logic.  I've moved forward on wl_seat, etc.  I can now drag the window between screens and the buffer scaling changes.  It's not to say the Mutter implementation tracks 100% correctly, but the nuance can be addressed at a later date if it shows up else where.  There was an issue when display scaling was > 100% it would segfault; without these changes.
